File: scripts/check-recent-scrapes.py
"""List leads written in the last N minutes. Verifies a recent scrape wrote to Firestore.

The Lead model timestamps with `scraped_at` (timezone-naive `datetime.now()`),
so we compare with a naive ISO string to match Firestore's lexicographic sort.

Run on the VPS (where Firebase creds are configured):
    cd /root/asterley-bros
    uv run python scripts/check-recent-scrapes.py            # default 60 min
    uv run python scripts/check-recent-scrapes.py 15         # last 15 min
"""

from __future__ import annotations

import sys
from datetime import datetime, timedelta

from dotenv import load_dotenv
load_dotenv()

from src.db.client import get_firestore_client


def main() -> int:
    minutes = int(sys.argv[1]) if len(sys.argv) > 1 else 60
    if minutes <= 0:
        print("Usage: uv run python scripts/check-recent-scrapes.py [minutes]")
        return 1

    db = get_firestore_client()
    if db is None:
        print("Firestore client unavailable. Check GOOGLE_APPLICATION_CREDENTIALS.")
        return 1

    # Naive UTC to match the format Pydantic writes (datetime.now() is naive).
    since_dt = datetime.utcnow() - timedelta(minutes=minutes)
    since = since_dt.isoformat()
    print(f"\n=== Leads scraped since {since} (last {minutes} min) ===\n")

    leads = list(db.collection("leads").where("scraped_at", ">=", since).stream())
    if not leads:
        print("No new leads in that window.")
    else:
        rows = sorted(
            (d.to_dict() for d in leads),
            key=lambda l: str(l.get("scraped_at", "")),
            reverse=True,
        )
        print(f"Found {len(rows)} lead(s):\n")
        for l in rows:
            scraped = str(l.get("scraped_at", ""))[:19].replace("T", " ")
            source = (l.get("source") or "?").ljust(14)
            where = l.get("website") or l.get("address") or "—"
            stage = l.get("stage") or "?"
            print(f"  {scraped}  [{source}] {l.get('business_name', '?')}")
            print(f"                       stage={stage}  {where}")

    print(f"\n=== Recent scrape_runs (last 10) ===\n")
    runs = (
        db.collection("scrape_runs")
        .order_by("started_at", direction="DESCENDING")
        .limit(10)
        .stream()
    )
    found_any = False
    for doc in runs:
        found_any = True
        r = doc.to_dict()
        started = str(r.get("started_at", ""))[:19].replace("T", " ")
        leads_found = r.get("leads_found", 0)
        status = r.get("status", "?")
        query = str(r.get("queries") or r.get("query") or "")[:40]
        print(f"  {started}  status={status}  leads={leads_found}  query=\"{query}\"")
    if not found_any:
        print("No scrape_runs found.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: functions/check-recent-scrapes.mjs
// List leads written in the last N minutes. Use this to verify a recent
// scrape actually wrote leads to Firestore.
//
// Queries `scraped_at` (not `created_at`) because that's the field every
// lead-create path actually writes. `created_at` is set on newer writes
// only; older docs lack it and Firestore would silently exclude them.
//
// Run from project root:
//   node functions/check-recent-scrapes.mjs            # default: last 60 min
//   node functions/check-recent-scrapes.mjs 15         # last 15 min
//   node functions/check-recent-scrapes.mjs 1440       # last 24 h
//   node functions/check-recent-scrapes.mjs 11520      # last 8 days

import { initializeApp, getApps, applicationDefault } from "firebase-admin/app";
import { getFirestore } from "firebase-admin/firestore";

const minutes = Number(process.argv[2] || 60);
if (!Number.isFinite(minutes) || minutes <= 0) {
  console.error("Usage: node check-recent-scrapes.mjs [minutes]");
  process.exit(1);
}

const app = getApps().length > 0
  ? getApps()[0]
  : initializeApp({ credential: applicationDefault(), projectId: "asterley-bros-b29c0" });
const db = getFirestore(app);

const since = new Date(Date.now() - minutes * 60 * 1000).toISOString();

console.log(`\n=== Leads written since ${since} (last ${minutes} min) ===\n`);

const snap = await db.collection("leads")
  .where("scraped_at", ">=", since)
  .get();

if (snap.empty) {
  console.log("No new leads in that window.");
} else {
  // Sort newest-first (Firestore where() doesn't include ordering here).
  const rows = snap.docs
    .map((d) => d.data())
    .sort((a, b) => String(b.scraped_at || "").localeCompare(String(a.scraped_at || "")));

  console.log(`Found ${rows.length} lead(s):\n`);
  for (const l of rows) {
    const ts = String(l.scraped_at || l.created_at || "").slice(0, 19).replace("T", " ");
    const source = l.source || "?";
    const stage = l.stage || "?";
    const where = l.website || l.address || "—";
    console.log(`  ${ts}  [${source.padEnd(14)}] ${l.business_name}`);
    console.log(`                       stage=${stage}  ${where}`);
  }
}

console.log(`\n=== Recent scrape_runs (last 10) ===\n`);
const runs = await db.collection("scrape_runs")
  .orderBy("started_at", "desc")
  .limit(10)
  .get();

if (runs.empty) {
  console.log("No scrape_runs found.");
} else {
  for (const doc of runs.docs) {
    const r = doc.data();
    const started = String(r.started_at || "").slice(0, 19).replace("T", " ");
    console.log(`  ${started}  status=${r.status || "?"}  leads=${r.leads_found ?? 0}  query="${(r.queries || r.query || "").toString().slice(0, 40)}"`);
  }
}
